machineconfiguration/v1alpha1/types_machineosbuild.go

```diff
@@ -0,0 +1,218 @@
+type ConfigMapData struct {
+	// key in the configmap pointing to the specific data
+	// +kubebuilder:validation:Required
+	Key string `json:"key"`
+	// value is the data stored in the configmap
+	// +optional
+	Value string `json:"value"`
+	// binaryValue is the data stored in the configmap
+	// +optional
+	BinaryValue []byte `json:"binaryValue"`
+}
+
+// Refers to the name of a MachineConfigPool (e.g., "worker", "infra", etc.):
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a rendered MachineConfig (e.g., "rendered-worker-ec40d2965ff81bce7cd7a7e82a680739", etc.):
+type RenderedMachineConfigReference struct {
+	// name is the name of the rendered MachineConfig object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of a (future) MachineOSImage (e.g., "worker-os-image-167651b10ec98af17971d6a47df9e22f", etc.):
+type MachineOSImageReference struct {
+	// name is the name of the MachineOSImage object attached to this MachineOSBuild.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// Refers to the name of an image registry push/pull secret needed in the build process.
+type ImageSecretObjectReference struct {
+	// name is the name of the secret used to push or pull this MachineOSBuild object.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// Valid options are: OpenShiftImageBuilder, PodImageBuilder, and Default (OpenShiftImageBuilder)
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+	// baseOSExtensionsImageURL is the base Extensions image used in the build process
+	// +kubebuilder:validation:Required
+	BaseOSExtensionsImageURL string `json:"baseOSExtensionsImageURL"`
+	// baseOSImageURL is the base OSImage we use to build our custom image.
+	// +kubebuilder:validation:Required
+	BaseOSImageURL string `json:"baseOSImageURL"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// +kubebuilder :validation:Required
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret"`
+	// finalImagePushSecret is the secret used to connect to a user registry.
+	// +kubebuilder:validation:Required
+	FinalImagePushSecret ImageSecretObjectReference `json:"finalImagePushSecret"`
+	// finalImagePullSecret is the secret used to pull the final produced image.
+	// +kubebuilder:validation:Required
+	FinalImagePullSecret ImageSecretObjectReference `json:"finalImagePullSecret"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// +optional
+	Containerfile string `json:"containerFile"`
```

Comment:
make this a set

Reply:
if this is a set how many  Containerfiles can you have before you hit 1mb resource limits?

Reply:
OpenShift supports a total of four different architectures (amd64, aarch64, ppc64le, s390x). Would someone actually build a cluster with all four architectures in the same cluster? Probably not, but it sounds like something I would do just to say that I did it :laughing:

That said, I'd say a cluster admin would only have two different architectures within the same cluster. But would adding a max of 4 cause excessive resource usage?

Reply:
As an aside, I have a question about the implementation detail here: How would we assign a Containerfile to a particular architecture? Would it make sense to do something like this?

```golang
type ContainerfileArch string

const (
	// Or we import them from here? https://github.com/openshift/api/blob/master/samples/v1/types_config.go#L178-L191
	// The only downside of doing that is they are just strings as opposed to a separate type.
	AMD64ContainerfileArch ContainerfileArch = "amd64"
	AARCH64ContainerfileArch ContainerfileArch = "aarch64"
	// ...
)

type Containerfile struct {
	Arch ContainerfileArch `json:"arch"`
	Contents string `json:"contents"`
}

// Then BuildInputs becomes something like:
type BuildInputs struct {
	// ...
	Containerfiles []Containerfile `json:"containerfiles"`
	// ...
}
```